### ISSUE TITLE
sun buttons have a different API for enabledProperty

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/sun#515]
Over in https://github.com/phetsims/sun/issues/514, this general issue was noted.   I found a workaround for that specific issue, but handling `enabledProperty` differently in buttons seems problematic.  And should be addressed when we do the "standardized how enabledProperty is handled" issue (which I can't seem to locate).

Here's my summary from https://github.com/phetsims/sun/issues/514#issuecomment-501400297:

Unlike other PhET UI components, sun buttons don't provide `options.enabledProperty` and they don't provide public `this.enabledProperty`. 

There is an `enabledProperty` internally, but it's in `ButtonModel`:

```js
    // @public - Is the button enabled?
    this.enabledProperty = new BooleanProperty( options.enabled, {
```

... but it's `@protected` in the view part of the button. For example in RectangularButtonView:

```js
    this.buttonModel = buttonModel; // @protected
```

And you're apparently supposed to call `setEnabled` and `getEnabled` to indirectly access `this.buttonModel.enabledProperty.set( value )`.

This is a problem in the case of AccordionBox, because the enabled state of its `ExpandCollapseButton` needs to be synchronized with whether clicking on the title bar expands/collapses the box.

**Comment by pixelzoom** [comment]
Assigning to @ariel-phet to prioritize and assign.

**Comment by pixelzoom** [comment]
Workaround for phetsims/graphing-quadratics#112 did not work, see https://github.com/phetsims/sun/issues/514#issuecomment-501407372.   In case that affects the priority...

**Comment by pixelzoom** [comment]
Addressing this doesn’t look easy, and probably requires discussion. My 2 cents is that this should be part of the bigger “standardize enabledProperty, maybe with a mixin” work. (Where is that pesky issue?)

**Comment by pixelzoom** [comment]
I said:
> ... (Where is that pesky issue?)

See https://github.com/phetsims/sun/issues/257#issuecomment-502900733 for the general issue of adding `enabledProperty` uniformly to UI components.

**Comment by samreid** [comment]
I think I understand the problem, and am curious as to whether this would block publication of GQIO 1.1.

**Comment by pixelzoom** [comment]
This doe not block GQIO 1.1, see https://github.com/phetsims/sun/issues/514#issuecomment-501877565.

**Comment by samreid** [comment]
On slack, @kathy-phet reminded us we would like to feature both `visibleProperty` and `enabledProperty` by default.  Hence it seems we need to have a good plan for `enabledProperty` for GQIO11.

**Comment by samreid** [comment]
Note this issue overlaps significantly with https://github.com/phetsims/sun/issues/257

**Comment by zepumph** [comment]
>Hence it seems we need to have a good plan for enabledProperty for GQIO11.

I understand that sentiment, though I'm not sure that we need to solve this issue for GQIO. This is mainly about a clunky private api. The enabledProperty interface for PhET-iO would stay the same I believe. I'm going to remove the label, let me know if you disagree.

**Comment by zepumph** [comment]
@pixelzoom cleared things up for me over in https://github.com/phetsims/sun/issues/514#issuecomment-506097034. 

Brainstorming here. What if Types like `RectangularButtonView` just had a getter for the enabledProperty, then we could link to it and do everything needed for https://github.com/phetsims/sun/issues/514.

**Comment by pixelzoom** [comment]
> Brainstorming here. What if Types like RectangularButtonView just had a getter for the enabledProperty, then we could link to it and do everything needed for #514.

That would be OK for a short-term solution, but will be more enabledProperty-related code that we'll need to revisit for https://github.com/phetsims/sun/issues/257#issuecomment-502900733.   If there's really a short-term need to do this, then I can't think of a better approach.

The longterm solution is to first decide what API we want to have to enabled/disable UI components over in https://github.com/phetsims/sun/issues/257#issuecomment-502900733, decide how to implement it uniformly (mixin?), then apply the API uniformly, including for sun buttons.  

> other UI components have a public enabledProperty, so I don't know why this one is so hidden.

I believe that sun buttons pre-date the boilerplate `enabledProperty` code used in other UI components.  Other UI components don't have the model/view separation that sun buttons have, and sun buttons have  `enabledProperty` in the (private) model.

**Comment by zepumph** [comment]
Above I added the two getters for the enabledProperty because I felt like this would be the easiest way to proceed with https://github.com/phetsims/sun/issues/514. I'm not really sure what else to do with this issue, because to me it feels like it should just be lumped in to work done in https://github.com/phetsims/sun/issues/257. For now I'm going to remove the blocking status and ask that @pixelzoom review the commit above with the fix for #514.

**Comment by pixelzoom** [comment]
If it were me, I would have just deferred #514, I don't enabledProperty for AccordionBox’s ExpandCollapseButton should be a priority for GQIO.  The use case for GQ (and I believe sims in general) is to hide the button so that students could not collapse the AccordionBoxes.  I see little value in disabling the button. 

That said... If you must address that issue, then the changes in https://github.com/phetsims/sun/commit/f05fe43bedd18951a727051a40280cfdb788f6f4 should be documented as temporary, with a link to this issue.  I'd also add a note discouraging developers from using, and explaining why.

**Comment by zepumph** [comment]
That seems fair enough to me. I removed an unused workaround in the round view, and added a TODO pointing to https://github.com/phetsims/sun/issues/257. @pixelzoom please close if that feels good to you.

**Comment by pixelzoom** [comment]
Thanks @zepumph, looks good.   Now that we have a temporary solution, I'm going to reassign this issue to @ariel-phet to prioritize and assign for a permanent solution.

**Comment by ariel-phet** [comment]
@zepumph perhaps we should have some tag/milestone or such that is some things we would like to address after GQIO is finished? It seems like then such issues should be reviewed and prioritized after GQIO is finished. This seems like one such issue. 

Marking for iO meeting mainly to discuss gathering and tagging relevant issues.

**Comment by chrisklus** [comment]
Since this relies on https://github.com/phetsims/sun/issues/257, @zepumph, @samreid, and I are assigning to @jbphet and marking on-hold. Once the work in #257 is complete, @jbphet please check if this was taken care of as part of #257, complete the remaining work here, or pass it back to us.

**Comment by pixelzoom** [comment]
We ran into this problem again over in https://github.com/phetsims/scenery-phet/issues/563, where we need to control _and observe_ the enabled state of `StepButton`.  Raising the priority and labeling for developer meeting.

**Comment by mattpen** [comment]
Discussed at 1/2 dev meeting: @jbphet will discuss priority of this after the current work on number-line is complete.

**Comment by pixelzoom** [comment]
More notes from 1/2/ dev meeting: @jbphet should check with @jessegreenberg about the priority of https://github.com/phetsims/scenery-phet/issues/563, since that issue is blocked until this is addressed.

**Comment by zepumph** [comment]
Self assigning as it pertains to work to be done in https://github.com/phetsims/sun/issues/257

**Comment by pixelzoom** [comment]
In the above commit, ButtonModel now uses mixin EnabledComponent.  This brings us a baby step forward, so that `enabledProperty` is now created correctly in ButtonModel.  And you can pass in `options.enabledProperty`.

But there are still 2 major differences with other UI components:

* The first difference is programmer API. You can't access `myButton.enabledProperty` because `enabledProperty` lives in the (protected) ButtonModel. So instead of `myButton.enabledProperty.value =...` you must do `myButton.enabled =...`

* The second difference is PhET-iO related. If you use `options.enabledProperty` to pass a Property to a button, that Property does not appear as a linked element in Studio.  That link is created by EnabledPhetioObject (a subclass of EnabledComponent), which can only be used if ButtonModel extends PhetioObject.  And we can't do that because it would mess up the Studio tree structure, exposing the ButtonModel.

**Comment by pixelzoom** [comment]
I don't know how to resolve the remaining `enabledProperty` differences described in https://github.com/phetsims/sun/issues/515#issuecomment-709625808.  Since the design and implementation of sun buttons has numerous other problems, you might consider a "sun buttons 2.0" implementation sometime in the not-too-distant future.  My recommendation there would be to either avoid the model-view separation that was attempted in the current implementation, or think it through more fully. That separation is what's responsible for the `enabledProperty` differences.

Back to @ariel-phet to decide how to proceed.

**Comment by zepumph** [comment]
I got here while reviewing https://github.com/phetsims/sun/issues/585.

I updated a TODO to point to this issue, unfortunately https://github.com/phetsims/sun/issues/257 didn't seem to solve this workaround. I agree that I don't see a way around this except to rethink how sun buttons work. We could do this on a micro scale for enabledProperty, but for that to make sense we would at least want to refactor the round/rectangular button views into a central supertype that we could use `EnabledNode` on (instead of using it in ButtonModel). This suggestion is not perfect, but was gathered from some conversation with @pixelzoom last week.

I too am skeptical and not confident about how to proceed on this issue.

**Comment by zepumph** [comment]
If we wanted to ignore the problem, this patch would obfuscate the button M/V separation, and potentially get us far in duplicating a consistent API for sun buttons:

```diff
Index: js/buttons/RectangularButtonView.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/buttons/RectangularButtonView.js	(revision adc57f6df12d74785e837f8db1fc71c6efb6ba12)
+++ js/buttons/RectangularButtonView.js	(date 1603313489575)
@@ -251,7 +251,7 @@
    * @returns {Property.<boolean>}
    * @public
    */
-  getEnabledProperty() { return this.buttonModel.enabledProperty; }
+  get enabledProperty() { return this.buttonModel.enabledProperty; }
 
   /**
    * Sets the base color, which is the main background fill color used for the button.

```

**Comment by pixelzoom** [comment]
The above patch solves difference (1) mentioned in https://github.com/phetsims/sun/issues/515#issuecomment-709625808.   Does it do anything to address difference (2), or can we come up a workaround?

**Comment by pixelzoom** [comment]
Self assigning. @zepumph's patch above has inspired me to take another stab at this.

**Comment by pixelzoom** [comment]
I added the patch suggested by @zepumph in https://github.com/phetsims/sun/issues/515#issuecomment-713870207.  I documented it as "intentional obfuscation":

```js
  /**
   * ES5 getter for the model's enabledProperty. This is a bit of intentional obfuscation to make sun buttons
   * have an enabledProperty API that is similar to other UI components.
   * See https://github.com/phetsims/sun/issues/515#issuecomment-713870207
   * @returns {Property.<boolean>}
   */
  get enabledProperty() { return this.getEnabledProperty(); }
```

And I tested it in ButtonsScreenView, where it makes the API look more typical:

```js
      // Test the enabledProperty API for these buttons, see https://github.com/phetsims/sun/issues/515
      buttonA.enabledProperty.value = enabled;
      buttonB.enabledProperty.value = enabled;
      buttonC.enabledProperty.value = enabled;
      buttonD.enabledProperty.value = enabled;
      buttonE.enabledProperty.value = enabled;
```

I feel conflicted after doing this, because it's a workaround that relies on obfuscation. But I don't have a better idea.  So... Should we keep this, or revert it?

**Comment by zepumph** [comment]
Best solution: fix sun buttons
other good solution: your commit

Probably we should keep it. As we want Button Node's api to act like all other enabledProperty apis, regardless of implementation. This will help future-proof us for when we actually make sun-buttons what we want.

**Comment by pixelzoom** [comment]
To fix problem (2), adding the linked element, add this to the constructors for RectangularButtonView and RoundButtonView:

```js
    //TODO https://github.com/phetsims/sun/issues/515 copied from EnabledPhetioObject
    if ( options.enabledProperty ) {
      if ( Tandem.VALIDATION && this.isPhetioInstrumented() ) {
        assert && assert( options.enabledProperty.isPhetioInstrumented(), 'provided enabledProperty must be instrumented if this PhetioObject is.' );
        assert && this.phetioFeatured && assert( options.enabledProperty.phetioFeatured, 'provided enabledProperty must be phetioFeatured if this PhetioObject is.' );
      }

      // If enabledProperty was passed in, PhET-iO wrappers like Studio needs to know about that linkage
      // This is supported in API.js types because in practice LinkedElementIO defer to their linked PhetioObject.
      this.addLinkedElement( options.enabledProperty, {
        tandem: options.tandem.createTandem( EnabledComponent.ENABLED_PROPERTY_TANDEM_NAME )
      } );
    }
```

I tested this in Natural Selection by adding `enabledProperty: new BooleanProperty( true, {...} )` to the ResetAllButton in NaturalSelectionScreenView.  It adds the linked element like this:

<img width="169" alt="screenshot_587" src="https://user-images.githubusercontent.com/3046552/96800987-de0ee280-13c3-11eb-8522-c2ae34a8bc27.png">

But as the comment says in the above patch, this involves copying code from EnabledPhetioObject.  And this (and the previous patch) also highlight a big flaw of sun buttons -- their design requires duplication in RectangularButtonView and RoundButtonView that has nothing to do with button shape.

@zepumph what do you think? Add this workaround too?

**Comment by zepumph** [comment]
I'm going to prototype some stuff and get back to you!

**Comment by zepumph** [comment]
Ok, above I did something weird, and I'm excited for feedback. I decided to add EnabledNode to ButtonNode (new as of https://github.com/phetsims/sun/issues/643), but instead of having it create its own, separate from the ButtonModel's, we just pass the model's enabledProperty in. This way we can keep the same Property reference, but add all the logic that the EnabledNode mixin subtype adds that we have been missing in sun buttons.

There is one large TODO still, as the linkedElement added in EnabledPhetioObject doesn't work yet. Thus, in a way we are no farther than before https://github.com/phetsims/sun/issues/515#issuecomment-713987551, still just working around the linked element addition, but in general we are still better off.

I hope that this will be fixed by having ButtonNode mutate options (and thus be in charge of instrumenting), so that we can initialize the EnabledNode after that. Please see https://github.com/phetsims/sun/issues/643#issuecomment-714093513 for more details on the issue though:

>ButtonNode doesn't mutate options, but instead it relies on subtypes to do it. This is because the sub-type views rely on calling mutate after the content has been set, as some formatting/layout relies on having content first. Even so, we may be able to refactor the subclasses to better support this, like perhaps passing up children or something.

**Comment by pixelzoom** [comment]
Re renaming `RectangularButtonView` and `RoundButtonView`....  Here are some of their subclasses:

```
class RectangularPushButton extends RectangularButtonView 
class RectangularToggleButton extends RectangularButtonView
...
class RoundPushButton extends RoundButtonView
class RoundToggleButton extends RoundButtonView
...
```

In this content, `RectangularButton` and `RectangularButton` seem like better names. With abstract base class `ButtonNode` or `Button` or `SunButton` or ... ?

**Comment by arouinfar** [comment]
On Slack, @kathy-phet said:
>  Can @arouinfar review the changes to button and enabledproperty and opacityproperty and visibleproperty as they are relevant to GAO before we go to RC-lite? They seem like they have been significant, so need designer review.

This comment seems at least partially related to the work being done in this issue. However, the work is still in progress and it is unclear what designer input is necessary. I'm tagging this issue for discussion in phet-io meeting.

**Comment by pixelzoom** [comment]
@zepumph said:
> Ok, above I did something weird, and I'm excited for feedback. I decided to add EnabledNode to ButtonNode (new as of #643), but instead of having it create its own, separate from the ButtonModel's, we just pass the model's enabledProperty in. ...

Yeah, a little weird, but I think it's OK.  It's certainly preferable to having to use something like DynamicProperty to sync 2 different enabledProperty instances for model and view.  In the above commit I cleanup up how this is done a tad.

**Comment by pixelzoom** [comment]
@zepumph said:
> ... There is one large TODO still, as the linkedElement added in EnabledPhetioObject doesn't work yet. Thus, in a way we are no farther than before #515 (comment), still just working around the linked element addition, but in general we are still better off.

Did you mean a literal TODO?  I don't see one related to this issue.

Thoughts on why `addLinkedElement` in EnabledPhetioObject isn't working (I haven't confirmed) and how to get it to work?

**Comment by zepumph** [comment]
The above commit fixes the problem as far as I can tell.

Further work over in https://github.com/phetsims/sun/issues/643 has made it easy to just initialize the EnabledNode after the mutate call. This required an extra check to make sure we don't try to add a LinkedElement in the `enabledProperty` spot when the actual Property has that phetioID.

I was able to remove the `phetioLinkEnabledElement` workaround. @pixelzoom, things are working for enabledProperty as far as I can tell. I did some fuzzing on a few sims, and poked around in NS studio. Can you please review this issue. Anything else here?

**Comment by pixelzoom** [comment]
I tested this in Natural Selection, with the ResetAllButton that is created in NaturalSelectionScreenView.

With no `enabledProperty` option to ResetAllButton, I see the expected tree structure in Studio:

<img width="157" alt="screenshot_622" src="https://user-images.githubusercontent.com/3046552/97618644-e7600680-19e4-11eb-99fc-4e86a97c5df2.png">

Then I added an `enabledProperty` options, like this:

```js
    const resetAllButton = new ResetAllButton( {
      enabledProperty: new BooleanProperty( true, {
        tandem: options.tandem.createTandem( 'resetAllButtonEnabledProperty' )
      } ),
      ...
      tandem: options.tandem.createTandem( 'resetAllButton' )
    } );
```

and I see the expected tree structure and linked element in Studio:

<img width="1000" alt="screenshot_625" src="https://user-images.githubusercontent.com/3046552/97619011-50e01500-19e5-11eb-9a1c-e2a0b2abff8b.png">

So this all looks great.

**Comment by pixelzoom** [comment]
Over to @arouinfar for PhET-iO design review.  She said:
> .... the work is still in progress and it is unclear what designer input is necessary.

Please confirm that:

(1) The tree structure is as expected, and similar to other PhET UI components. That is, the button model is not exposed, and `enabledProperty` is directly under the UI component.

(2) If an `enabledComponent` is supplied to the button, the tree structure should be the same, with the addition of an element named `enabledProperty` that is linked to the supplies enabledProperty.

**Comment by pixelzoom** [comment]
This is a prerequisite for moving forward with some dev (Natural Selection) and RC versions.  So increasing the priority of PhET-iO design to high.

**Comment by pixelzoom** [comment]
Discussed at 10/29/2020 phet-io meeting. @arouinfar if you have any more questions about what/how to review the enabledProperty PhET-iO API for sun buttons, let me know.

This is a prerequisite for moving forward with some dev testing (Natural Selection) and RC testing. So increasing the priority of PhET-iO design to high.

**Comment by pixelzoom** [comment]
I see in https://github.com/phetsims/scenery/issues/1100 that work is not done on enabledProperty.  It's now moving into Node.  That work has the potential (though unlikely) to introduce regressions in the PhET-iO API, so it probably makes sense to have @arouinfar wait to review `enabledProperty` in sun buttons until that work has been completed. 

Assigning to @zepumph and @samreid, because I don't know who is the lead on  https://github.com/phetsims/scenery/issues/1100. When that work is completed, please assign this issue back to @arouinfar for review.

**Comment by zepumph** [comment]
Thanks @pixelzoom. I kept @arouinfar apprised privately via slack on Friday but didn't update this issue. That work is complete in regards to it blocking this issue. This is still ready for review.

**Comment by arouinfar** [comment]
I reviewed this in the context of Gravity and Orbits, and compared master to the most recently QA-tested version, [1.2.0-dev.16](https://phet-dev.colorado.edu/html/gravity-and-orbits/1.2.0-dev.16/phet-io/wrappers/studio/). Generally, things are looking really nice and the tree seems undisturbed. The only difference I noticed was that setting the `enabledProperty` of the buttons in the navbar false doesn't remove the hover behavior. The vertical lines appear on either side of the button even though it is not pickable. This did not occur previously, and I think the hover behavior creates the illusion that the buttons are still enabled. Affected phetioIDs:
* `gravityAndOrbits.general.view.navigationBar.homeButton.enabledProperty`
* `gravityAndOrbits.general.view.navigationBar.modelScreenButton.enabledProperty`
* `gravityAndOrbits.general.view.navigationBar.toScaleScreenButton.enabledProperty`
* `gravityAndOrbits.general.view.navigationBar.phetButton.enabledProperty`

**Comment by pixelzoom** [comment]
Re navigation bar buttons...

**Comment by pixelzoom** [comment]
Re the problems with navigation bar buttons reported by @arouinfar in https://github.com/phetsims/sun/issues/515#issuecomment-720709179...  Navigation bar buttons are not sun buttons. Joist has it's own button implementation.  So please see https://github.com/phetsims/joist/issues/676.

**Comment by zepumph** [comment]
Good find @arouinfar. That is awesome to see. 

Thanks @pixelzoom for creating the issue. I'll take the lead on that.

>Generally, things are looking really nice and the tree seems undisturbed.

From this, I feel like we are ready to close this issue. Anyone, feel free to reopen if there is anything else.
